apps/src/netsim/NetSimVisualization.js

```diff
@@ -0,0 +1,475 @@
+    } else {
+      vizNode = new NetSimVizNode(node);
+      vizNode.snapToPosition(
+          Math.random() * this.visualizationWidth - (this.visualizationWidth / 2),
+          Math.random() * this.visualizationHeight - (this.visualizationHeight / 2));
+      this.entities_.push(vizNode);
+      this.svgRoot_.find('#background_group').prepend(vizNode.getRoot());
+    }
+  }, this);
+};
+
+/**
+ * Handle notification taht wire table contents have changed.
+ * @param {Array.<Object>} rows - wire table rows
+ * @private
+ */
+NetSimVisualization.prototype.onWireTableChange_ = function (rows) {
+  // Convert rows to correctly-typed objects
+  var tableWires = rows.map(function (row) {
+    return new NetSimWire(this.shard_, row);
+  }.bind(this));
+
+  // 1. Kill wires that are no longer in the table
+  this.entities_.filter(function (vizEntity) {
+    return vizEntity instanceof NetSimVizWire;
+  }).filter(function (vizWire) {
+    return !tableWires.some(function (wire) {
+      return wire.entityID === vizWire.id;
+    });
+  }).forEach(function (vizWire) {
+    vizWire.kill();
+  });
+
+  // 2. Add new wires into the visualization
+  var vizWire;
+  tableWires.forEach(function (wire) {
+    vizWire = this.getEntityByID(NetSimVizWire, wire.entityID);
+    if (vizWire) {
+      vizWire.configureFrom(wire);
+    } else {
+      vizWire = new NetSimVizWire(wire, this.getEntityByID.bind(this));
+      this.entities_.push(vizWire);
+      this.svgRoot_.find('#background_group').prepend(vizWire.getRoot());
+    }
+  }, this);
+
+  // Since the wires table determines simulated connectivity, we trigger a
+  // recalculation of which nodes are in the local network (should be in the
+  // foreground) and then re-layout the foreground nodes.
+  this.pullElementsToForeground();
+  this.distributeForegroundNodes();
+};
+
+/**
+ * Recalculate which nodes should be in the foreground layer by doing a full
+ * traversal starting with the local node.  In short, everything reachable
+ * from the local node belongs in the foreground.
+ */
+NetSimVisualization.prototype.pullElementsToForeground = function () {
+  // Begin by marking all entities background (unvisited)
+  this.entities_.forEach(function (vizEntity) {
+    vizEntity.speculativeIsForeground = false;
+  });
+
+  // Use a simple stack for our list of nodes that need visiting.
+  // If we have a local node, push it onto the stack as our starting point.
+  // (If we don't have a local node, the next step is REALLY EASY)
+  var exploreStack = [];
+  if (this.localNode) {
+    exploreStack.push(this.localNode);
+  }
+
+  // While there are still nodes that need visiting,
+  // visit the next node, marking it as "foreground/visited" and
+  // pushing all of its unvisited connections onto the stack.
+  var currentVizEntity;
+  while (exploreStack.length > 0) {
+    currentVizEntity = exploreStack.pop();
+    this.visitEntityToSetForeground_(currentVizEntity, exploreStack);
+  }
```

Comment:
The method `visitEntityToSetForeground` puts things on the stack, but I agree that isn't obvious here.  I'll figure out a different breakdown so that all stack operations show up together.